analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs

```diff
@@ -287,5 +288,246 @@ public void DeclarationPattern_Var_Deconstruction_DoesNotSetConstraint()
+        [DataRow("stringUnknown", "not null or { Length: 0 }", null)]
+        [DataRow("stringNull", "{ Length: > 10 } or { Length: < 100 }", false)]
+        [DataRow("stringNotNull", "{ Length: > 10 } or { Length: < 100 }", null)]
+        [DataRow("stringUnknown", "{ Length: > 10 } or { Length: < 100 }", null)]
+        public void AndOrPatternsSetBoolConstraint(string variableName, string isPattern, bool? expectedBoolConstraint)
+        {
+            const string variableDeclarations = @"
+var objectNotNull = new object();
+var objectNull = (object)null;
+var objectUnknown = Unknown<object>();
+var stringNotNull = new string('c', 1);  // Make sure, we learn 'stringNotNull is not null'
+var stringNull = (string)null;
+var stringUnknown = Unknown<string>();
+var exceptionNotNull = new Exception();
+";
+            ValidateSetBoolConstraint(variableDeclarations, variableName, isPattern, OperationKindEx.BinaryPattern, expectedBoolConstraint);
+        }
+
+        private static void ValidateSetBoolConstraint(string variableDeclarations,
+                                                      string variableName,
+                                                      string isPattern,
+                                                      OperationKind expectedOperation,
+                                                      bool? expectedBoolConstraint)
+            => ValidateSetBoolConstraint(additionalTypes: string.Empty, variableDeclarations, variableName, isPattern, expectedOperation, expectedBoolConstraint);
+
+        private static void ValidateSetBoolConstraint(string additionalTypes,
+                                                      string variableDeclarations,
+                                                      string variableName,
+                                                      string isPattern,
+                                                      OperationKind expectedOperation,
+                                                      bool? expectedBoolConstraint)
+        {
+            var code = @$"
+{additionalTypes}
+public void Main()
+{{
+    {variableDeclarations}
+
+    var result = {variableName} is {isPattern};
+    Tag(""Result"", result);
+}}";
+            var validator = SETestContext.CreateCSMethod(code).Validator;
+            validator.ValidateContainsOperation(expectedOperation);
+            validator.ValidateTag("Result", x =>
+            {
+                if (expectedBoolConstraint is bool expected)
+                {
+                    x.Should().NotBeNull("we expect an constraint on the symbolValue");
+                    x.HasConstraint(BoolConstraint.From(expected)).Should().BeTrue("we should have learned that result is {0}", expected);
+                }
+                else
+                {
+                    if (x != null)
+                    {
+                        x.HasConstraint<BoolConstraint>().Should().BeFalse("we should not learn about the state of result");
```

Comment:
Why not `else x.Should().BeNull(...)`? This tweak is counter intuitive for the `expectedBoolConstraint`

Reply:
There is no "Or" for fluent validations. It is okay if `x is null || x.HasConstraint<Bool>() == false`.

Reply:
You're asserting `null or doesn't have Bool`. Where `null` also means `doesn't have Bool` as a consequence.

Let's assert .Should().BeNull() instead explicitly